server/data/data-cassandra/src/main/java/org/apache/james/rrt/cassandra/CassandraMappingsSourcesDAO.java

```diff
@@ -0,0 +1,108 @@
+import com.datastax.driver.core.Session;
+import com.github.steveash.guavate.Guavate;
+
+import reactor.core.publisher.Mono;
+
+class CassandraMappingsSourcesDAO {
+    private final CassandraAsyncExecutor executor;
+    private final CassandraUtils cassandraUtils;
+    private final PreparedStatement insertStatement;
+    private final PreparedStatement deleteStatement;
+    private final PreparedStatement retrieveSourcesStatement;
+
+    @Inject
+    CassandraMappingsSourcesDAO(Session session, CassandraUtils cassandraUtils) {
+        this.executor = new CassandraAsyncExecutor(session);
+        this.cassandraUtils = cassandraUtils;
+        this.insertStatement = prepareInsertStatement(session);
+        this.deleteStatement = prepareDelete(session);
+        this.retrieveSourcesStatement = prepareRetrieveSourcesStatement(session);
+    }
+
+    private PreparedStatement prepareInsertStatement(Session session) {
+        return session.prepare(insertInto(TABLE_NAME)
+            .value(MAPPING_TYPE, bindMarker(MAPPING_TYPE))
+            .value(MAPPING_VALUE, bindMarker(MAPPING_VALUE))
+            .value(SOURCE, bindMarker(SOURCE)));
+    }
+
+    private PreparedStatement prepareDelete(Session session) {
+        return session.prepare(delete()
+            .from(TABLE_NAME)
+            .where(eq(MAPPING_TYPE, bindMarker(MAPPING_TYPE)))
+            .and(eq(MAPPING_VALUE, bindMarker(MAPPING_VALUE)))
+            .and(eq(SOURCE, bindMarker(SOURCE))));
+    }
+
+    private PreparedStatement prepareRetrieveSourcesStatement(Session session) {
+        return session.prepare(select(SOURCE)
+            .from(TABLE_NAME)
+            .where(eq(MAPPING_TYPE, bindMarker(MAPPING_TYPE)))
+            .and(eq(MAPPING_VALUE, bindMarker(MAPPING_VALUE))));
+    }
+
+    Mono<Void> addMapping(Mapping mapping, MappingSource source) {
+        return executor.executeVoidReactor(insertStatement.bind()
+            .setString(MAPPING_TYPE, mapping.getType().asPrefix())
+            .setString(MAPPING_VALUE, mapping.getMapping())
+            .setString(SOURCE, source.asMailAddressString()));
+    }
+
+    Mono<Void> removeMapping(Mapping mapping, MappingSource source) {
+        return executor.executeVoidReactor(deleteStatement.bind()
+            .setString(MAPPING_TYPE, mapping.getType().asPrefix())
+            .setString(MAPPING_VALUE, mapping.getMapping())
+            .setString(SOURCE, source.asMailAddressString()));
+    }
+
+    Mono<List<MappingSource>> retrieveSources(Mapping mapping) {
+        return executor.executeReactor(retrieveSourcesStatement.bind()
+            .setString(MAPPING_TYPE, mapping.getType().asPrefix())
+            .setString(MAPPING_VALUE, mapping.getMapping()))
+            .map(resultSet -> cassandraUtils.convertToStream(resultSet)
+                .map(row -> row.getString(SOURCE))
+                .map(MappingSource::parse)
+                .collect(Guavate.toImmutableList()));
```

Comment:
`ImmutableList`

Reply:
No, as always, we have to keep using Guavate as the feature is experimental in Guava cf @mbaechler

Reply:
No List anymore anyway, switched to a `Flux` !